### ISSUE TITLE
[RFC][TwigBundle] Register custom functions & filters using an attribute

### ISSUE THREAD

**Issue by artyuum** [symfony/symfony#50016]
### Description

Currently, in order to create a custom twig function/filter, we need create a class that extends the AbstractExtension, that looks like this:

```php
class TwigExtension extends AbstractExtension
{
    public function getFilters(): array
    {
        return [
            new TwigFilter('my_custom_filter', 'myCustomFilter'),
        ];
    }

    public function getFunctions(): array
    {
        return [
            new TwigFunction('my_custom_function', 'myCustomFunction'),
        ];
    }

    public function myCustomFilter(): string
    {
        // code here
    }

    public function myCustomFunction(): string
    {
        // code here
    }
}
```

### Example

I've always thought that the DX regarding twig extension could be better when working with Symfony. Today, I imagined the following:
```php
class TwigExtension extends AbstractExtension
{
    #[TwigFilter('my_custom_filter')]
    public function myCustomFilter(): string
    {
        // code here
    }

    #[TwigFunction('my_custom_function')]
    public function myCustomFunction(): string
    {
        // code here
    }
}
```

If we don't want to modify or conflict with the existing `TwigFunction` / `TwigFilter` classes, we could name them like this and store them in the twig bridge I think?
- `#[AsTwigFilter]`
- `#[AsTwigFunction]`

**What do you think?**

~~(If the feedbacks are positive, I'll open a PR.)~~ Hum, this seems more difficult than I initially thought.

**Comment by evertharmeling** [comment]
I guess you could even go as far as even deprecating the `AbstractExtension/ExtensionInterface`. And maybe introduce an `#[AsTwigExtension]` for the class (to narrow the classes to search for the attributes). But that could be a separate PR.

Also be sure to include the `TokenParserInterface`, `NodeVisitorInterface`, `Operator`, `TwigTest` as well, so full list of attributes would become:
- `#[AsTwigFilter]`
- `#[AsTwigFunction]`
- `#[AsTwigTest]`
- `#[AsTwigOperator]`
- `#[AsTwigTokenParser]` (for consistency include the `Twig` prefix?)
- `#[AsTwigNodeVisitor]` (for consistency include the `Twig` prefix?)

**Comment by stof** [comment]
Be careful that Twig itself cannot deprecate the existing system if the new system cannot easily work when using Twig outside Symfony.

And making TwigBundle expose a Symfony-only way of extending Twig might not be the best way to go. Especially when you will still need to go back to the Twig way to use some of the features as an attribute won't be able to configure the `is_safe_callback` option of Twig filters or functions as you cannot put a closure in an attribute)

> I guess you could even go as far as even deprecating the `AbstractExtension/ExtensionInterface`. And maybe introduce an `#[AsTwigExtension]` for the class (to narrow the classes to search for the attributes). But that could be a separate PR.

Twig uses the extension instance. So ExtensionInterface makes sense to define its API. And I doubt we can deprecate that without breaking support for standalone usage of Twig.

**Comment by artyuum** [comment]
I agree that deprecating these interfaces wouldn't make sense since they are not defined nor only used in the framework but are part of the twig package itself and they are still "valid".

> Be careful that Twig itself cannot deprecate the existing system if the new system cannot easily work when using Twig outside Symfony.

  I was thinking of implementing that only in the framework (either in the TwigBridge or the TwigBundle). 

> And making TwigBundle expose a Symfony-only way of extending Twig might not be the best way to go. Especially when you will still need to go back to the Twig way to use some of the features as an attribute won't be able to configure the is_safe_callback option of Twig filters or functions as you cannot put a closure in an attribute)

Would it be acceptable to still offer these new attributes as an easier/faster alternative for registering the custom functions/filters when using the framework but still suggesting the Twig way when doing something more complex that wouldn't be compatible with the attribute (e.g. callback)?

**Comment by SVillette** [comment]
I really love the idea. I only wonder if it is going possible to use the same attribute on a lazy extension. I mean, in your example, you declare the Twig function in the same class as its implementation (in a subclass of `AbstractExtension`). In case of lazy loaded extensions, the function/filter is declared in the extension which references a twig runtime (in a class that implements `RuntimeExtensionInterface`).
https://symfony.com/doc/current/templates.html#creating-lazy-loaded-twig-extensions

So my question is (and sorry if it is a dumb one), if you use attributes, how it will create an extension rather than a lazy loaded one ?

**Comment by stof** [comment]
Well, what we could do in Symfony is have 1 extension implemented in TwigBundle that gets configured based on those services using those attributes (those services will be registered as runtime).
RuntimeExtensionInterface is purely a marker interface to make it easy to identify classes that should be registered as Twig runtimes, but there is no requirement to actually use it.

**Comment by GromNaN** [comment]
I thought of an implementation in Twig directly, making the `getFilters`, `getFunctions` and `getTests` methods scan the attributes of the methods in the current class twigphp/twig#3916
